feat(http-log): add allow list for logged headers

ItsKaleb opened the pull request:
For some use cases, only a set of known, allowed headers should be included in traffic logs. This updates the http-log plugin to add an optional behaviour to only include a set of allowed request/response headers in the generated traffic logs.

<!--
NOTE: Please read the CONTRIBUTING.md guidelines before submitting your patch,
and ensure you followed them all:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributing
-->

### Summary

This PR offers an alternative default behaviour for the http-log plugin. Some use cases may require only specific headers to be logged for requests/responses. This may reduce risk of end-user injecting garbage to a header, or can restrict sensitive information from being logged from a non-standard header.

By adding the provided array of header names, this PR supplements the capabilities of the custom_fields config section. It provides an easier way to declare an allow list of logged headers, instead of needing to define a custom field rule for each header (request and response).

Additionally, by providing a default value of common headers, this allows the functionality to be on-click-opt-in using the separate enable toggle.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [x] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - https://github.com/Kong/docs.konghq.com/pull/5984

### Full changelog

* Add new fields to schema
* Implement logic to enforce only logging headers that appear in the allow list
* Add unit test

Note: While an effort was made to validate this PR, due to issues experienced with running a Dev Container on a fresh clone (see issue #11422 ) I have been unable to run the test suite. I have verified that the logical implementation does work by means of testing on a Kong 2.8.4.0 environment using a duplicate of the http-log plugin. Once issue #11422 is addressed, I am happy to confirm with further testing.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=11427) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=11427) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=11427) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=11427) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=11427) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=11427) it.</sub>

ItsKaleb commented:
Hi, 
My apologies if I've missed anything in raising this PR. I've tried to go through the appropriate contributing documentation. 
I was hoping to have this functionality added for Kong 2.8, though with the dev needing to be forked form the master branch, I wasn't sure on how to do the back-porting. As mentioned in the PR comment, I originally tested this functionality out for 2.8.4.0. I'm happy to update the PR for the backport if anything is needed - I might just need a point in the right direction, please.

Thanks

hanshuebner commented on kong/plugins/http-log/handler.lua:
```diff
@@ -181,13 +181,39 @@ end
+    local request_get_header = kong.request.get_header
+    local response_get_header = kong.response.get_header
+    for _, header_name in ipairs(conf.logged_headers_allow_list) do
+      local request_header_val = request_get_header(header_name)
+      local response_header_val = response_get_header(header_name)
+
+      if request_header_val then
+        set_serialize_value("request.headers." .. header_name, request_header_val)
+      end
+
+      if response_header_val then
+        set_serialize_value("response.headers." .. header_name, response_header_val)
+      end
```
```suggestion
      local request_header_val = request_get_header(header_name)
      if request_header_val then
        set_serialize_value("request.headers." .. header_name, request_header_val)
      end

      local response_header_val = response_get_header(header_name)
      if response_header_val then
        set_serialize_value("response.headers." .. header_name, response_header_val)
      end
```

hanshuebner reviewed (changes requested):
Hello @ItsKaleb,

thank you for your contribution!  The feature seems sensible.  A number of additional things is required, though, see below.  Note that once the PR is complete and passes testing, it will become part of the Kong 3.5 release train.  At this point, new features are not backported to 2.8.x.

Please add a [database migration](https://github.com/Kong/kong/tree/master/kong/plugins/http-log/migrations) - Remember to add it to the init.lua and to [rockspec file](https://github.com/Kong/kong/blob/master/kong-3.5.0-0.rockspec) as well.  For the migration, a [test](https://github.com/Kong/kong/tree/master/spec/05-migration/plugins/http-log/migrations) is needed.  Finally, the new fields need to be added to the [removed_fields.lua](https://github.com/Kong/kong/blob/master/kong/clustering/compat/removed_fields.lua) file to support mixed-version clusters.

ItsKaleb commented:
Thanks for the feedback, @hanshuebner. I will follow up on those points.

ItsKaleb commented:
Regarding the back port, can I raise a second PR targeting the release/2.8 branch?

hanshuebner commented:
> Regarding the back port, can I raise a second PR targeting the release/2.8 branch?

Yes, but we can't give you a release date for the next 2.8.x minor release.

hanshuebner commented:
> > Regarding the back port, can I raise a second PR targeting the release/2.8 branch?
> 
> Yes, but we can't give you a release date for the next 2.8.x minor release.

I need to back paddle regarding this:  It is unlikely that there will be another 2.8.x release made by us.  The reason for that is that we no longer have the release building infrastructure in place that was used before the 3.x release came out.  It also has been our policy not to backport features to old open source releases of Kong.  We do maintain 2.8.x for customers of the  Enterprise Edition (EE), but feature backports in EE require upper-level management approval, so there is no automatism for this.

If you cannot upgrade to Kong 3.5, you still have the option of forking, but you will need to invest in build infrastructure so that you can build whatever deployment artifacts that you require.

From the perspective of easy building, Kong 3.x with its Bazel based build system is much improved and makes it much easier to set up a separate build chain if you cannot directly use our releases for whatever reason.

ItsKaleb commented:
Yeah, I understand. 

My organisation does have a platinum license for EE. We should be upgrading to 3.5 soon, but I don't have a definite timeline. 

I've proven out a workaround to achieve the same functionality using the custom fields by lua config, it's just really clunky. 

Up until this point, we've used a custom logging plugin that was based on an older version of the http-log plugin; we had the logged header allow-list feature as part of that. Over time though, that plugin has developed some issues and I'm trying to migrate things back to the built-in option. 

Based on the info you've provided, what I'll do is continue rounding out this PR to add the feature for 3.5. I can either continue with the workaround in 2.8, or worst case scenario I already know I can duplicate the plugin from your source and add the feature we need.

ItsKaleb commented:
Hi @hanshuebner, based on the comments from yourself and @ms2008, I've updated the removed_fields.lua
I've just added a section for 3.5.x based on the existing patterns, please let me know if I've missed something.

As for the migrations, I just wanted to check if that was still required, or what would be needed for that. 
I've had a look through DEVELOPER.md and the existing migration for http-log - not sure if there'd be anything required for adding the new fields or not.
Are you able to confirm or point me in the right direction with that one?

Thanks

hanshuebner commented:
@ms2008 was completely right, no migration is needed because you're only adding new fields that have defaults.  Your change to `removed_fields.lua` looks good.  I'm going to let the tests run and will merge the PR if everything passes.

hanshuebner commented:
@ItsKaleb Can you address the test failures as far as they're related to your changes?  Thank you!

hanshuebner commented:
@vm-001 Can you please add a CHANGELOG file?

vm-001 commented on kong/plugins/http-log/schema.lua:
```diff
@@ -42,6 +42,43 @@ return {
+          { logged_headers_allow_list = {
+            type = "array",
+            elements = typedefs.header_name,
+            default = {
```
I think this list doesn't cover all the standard HTTP headers, not even including the most commonly used set, `Accept-Charset`, `Connection`, and `Keep-Alive` for example.

The standard(registered) HTTP header list can be found at https://datatracker.ietf.org/doc/html/rfc4229.

There are some non-standard(registered) headers in this list(ipremoteaddr, ratelimit-*, x-original-*, x-request-start), these don't work on other users.

Regarding this list, I have a concern. CP(with this feature) will send all the headers to the DP(with this feature), if users provision a lot of`http-log` plugin instances, the configuration size may increase significantly, which may put too much pressure on the DP memory, since DP needs load all the configuration in the memory first(correct me if I'm wrong).

Another way is to define a separated list. For example

```lua
-- constants.lua under http-log plugin
local M = {
  registered_headers = { "Accept", "Accept-Encoding", ... } 
}

return M
```
-- schema.lua
{ logged_headers_allow_list = { type = "array", ..., default = { "registered_headers" } } }
-- ...
```
```lua

@hanshuebner Thought?

hanshuebner replied:
@vm-001 Just to understand the concern:  The Control Plane will send a completely expanded configuration to the Data Plane, with all defaults filled in. This means that even if the feature is not used, all instances of the `http-log` plugin in the configuration will have the list of allowed headers attached.  If I'm understanding this correctly, then your concern is very valid.  I do not think, however, that a compiled-in list of standard headers is the right way to go.  Instead, I would propose that no default for `logged_headers_allow_list` is specified and that it would be an error to enable `enable_logged_headers_allow_list` without specifying `logged_headers_allow_list` as well.

vm-001 replied:
No default value for logged_headers_allow_list should address my concern.

ItsKaleb replied:
A default list for common headers was just a ease of use idea, I'm happy to apply the suggested changes of having no default list if that's the preferred approach.

ItsKaleb replied:
I've updated the default value to an empty array.

vm-001 commented on kong/plugins/http-log/schema.lua:
```diff
@@ -42,6 +42,43 @@ return {
           }},
           { queue = typedefs.queue },
           { custom_fields_by_lua = typedefs.lua_code },
+          { enable_logged_headers_allow_list = {
```
Either rename `enable_logged_headers_allow_list` to `enable_logged_header_allow_list` or vice versa in the specs.

ItsKaleb replied:
Ah yep, I see the typos now. Should be headers with 's' across the whole PR. Will address, thanks.

ItsKaleb replied:
@vm-001, this should be sorted now.

vm-001 reviewed (changes requested):
Test failed.

vm-001 commented:
> @vm-001 Can you please add a CHANGELOG file?

Done

hanshuebner commented:
> > @vm-001 Can you please add a CHANGELOG file?
> 
> Done

Thank you, also for your diligent review 💪

ItsKaleb commented:
Thanks both for your input. 

Regarding failing tests, I will be reviewing soon.

ItsKaleb commented:
Just to summarise my recent commits, I have addressed some typos that were highlighted in the new variable naming, as well as resolving a few suspect issues with the added tests. I believe things should be running smoother now.

ItsKaleb commented:
Hi @hanshuebner, 
I think I've managed to get my added tests sorted with this last PR, but there still seems to be a few unit tests failing. I'm not entirely sure if these are because of the changes I've made or not - they look to be related to the hybrid schema. Would you mind taking a look at the output please?

hanshuebner commented:
@ItsKaleb The failing process_auto_fields test needs to be updated so that the new fields are present in the expected responses (this is in line with the purpose of the test and cannot be avoided).  The asterisk in the error output tells you where the mismatch in the two structures is.  If you have a local build, you should be able to run just the failing test(s) using the `bin/busted` script which is in the repository.

ItsKaleb commented:
Thanks, will have another look with this new info.
I'm still struggling to get a local environment set up, but that's mainly to do with the fact I'm still trying to use the dev containers - I added some comments to issue #11422, just haven't had a whole lot of free time to tackle these two.
